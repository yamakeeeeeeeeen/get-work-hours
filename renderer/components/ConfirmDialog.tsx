import { FC, memo } from 'react';
import { Box, Button, Dialog, Typography } from '@material-ui/core';

type Props = {
  isOpen: boolean;
  setClose: () => void;
  body: string;
  callback: () => void;
};

export const ConfirmDialog: FC<Props> = memo(({ isOpen, setClose, body, callback }) => {
  return (
    <>
      <Dialog open={isOpen}>
        <Typography variant="h4">{body}</Typography>
        <Box display="flex">
          <Button variant="contained" color="primary" onClick={callback}>
            Yes
          </Button>
          <Button variant="contained" color="secondary" onClick={setClose}>
            No
          </Button>
        </Box>
      </Dialog>
    </>
  );
});

ConfirmDialog.displayName = 'ConfirmDialog';
