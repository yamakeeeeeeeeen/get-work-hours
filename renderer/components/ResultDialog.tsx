import { FC, memo } from 'react';
import { Box, Button, Dialog, Typography } from '@material-ui/core';

type Props = {
  isOpen: boolean;
  setClose: () => void;
  result: { breakTime: string; uptime: string };
};

export const ResultDialog: FC<Props> = memo(({ isOpen, setClose, result }) => {
  return (
    <>
      <Dialog open={isOpen}>
        <Typography variant="h4">{result.uptime}</Typography>
        <Typography variant="h4">{result.breakTime}</Typography>
        <Box display="flex">
          <Button variant="contained" color="secondary" onClick={setClose}>
            CLOSE
          </Button>
        </Box>
      </Dialog>
    </>
  );
});

ResultDialog.displayName = 'ResultDialog';
