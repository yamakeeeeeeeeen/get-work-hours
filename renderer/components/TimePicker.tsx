import { createStyles, makeStyles, TextField, Theme } from '@material-ui/core';
import { FC } from 'react';

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    textField: {
      marginLeft: theme.spacing(1),
      marginRight: theme.spacing(1),
      width: 120,
    },
  }),
);

type Props = {
  label: string;
};

export const TimePicker: FC<Props> = ({ label }) => {
  const classes = useStyles();

  return (
    <TextField
      label={label}
      type="time"
      defaultValue="12:00"
      variant="outlined"
      color="primary"
      className={classes.textField}
      InputLabelProps={{
        shrink: true,
      }}
    />
  );
};
