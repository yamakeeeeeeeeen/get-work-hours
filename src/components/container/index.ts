export { Header } from '~/components/container/Header';
