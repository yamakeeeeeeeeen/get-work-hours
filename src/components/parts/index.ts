export { AppBar } from '~/components/parts/AppBar';
export { Box } from '~/components/parts/Box';
export { Button } from '~/components/parts/Button';
export { Toolbar } from '~/components/parts/Toolbar';
export { Typography } from '~/components/parts/Typography';
