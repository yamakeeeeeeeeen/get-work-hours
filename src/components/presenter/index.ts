export { Header } from '~/components/presenter/Header';
