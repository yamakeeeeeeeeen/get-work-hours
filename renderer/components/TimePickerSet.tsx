import { FC, memo } from 'react';
import { UseFormMethods } from 'react-hook-form';
import { Box, Button, Typography } from '@material-ui/core';
import { Props } from '~/containers/TimePickerSet';
import { TimePicker } from '~/components/TimePicker';
import { ConfirmDialog } from '~/components/ConfirmDialog';
import { getValidation } from '~/utils/getValidation';
import { ORDINAL_NUMBERS } from '~/constants';

type ComponentProps = Omit<Props, 'fields' | 'remove'> & {
  isDisabled: boolean;
  isOpen: boolean;
  setOpen: () => void;
  setClose: () => void;
  errors: UseFormMethods['errors'];
  getValues: UseFormMethods['getValues'];
  register: UseFormMethods['register'];
  handleRemove: () => void;
};

export const Component: FC<ComponentProps> = memo(
  ({ index, value, isDisabled, isOpen, setOpen, setClose, errors, getValues, register, handleRemove }) => {
    const { biggerThanThePrevious, formatHHMM } = getValidation();

    return (
      <>
        <Box display="flex" alignItems="center" mb={3}>
          <Typography>{ORDINAL_NUMBERS[index + 1]}</Typography>
          <TimePicker
            name={`workTimes[${index}].start`}
            label="start"
            defaultValue={value.start}
            inputRef={register({
              validate: {
                ...formatHHMM(),
                ...biggerThanThePrevious('start', index, getValues),
              },
            })}
            error={!!errors?.workTimes?.[index]?.start}
            helperText={errors?.workTimes?.[index]?.start && errors?.workTimes?.[index]?.start?.message}
          />
          <Typography>~</Typography>
          <TimePicker
            name={`workTimes[${index}].end`}
            label="end"
            defaultValue={value.end}
            inputRef={register({
              validate: {
                ...formatHHMM(),
                ...biggerThanThePrevious('end', index, getValues),
              },
            })}
            error={!!errors?.workTimes?.[index]?.end}
            helperText={errors?.workTimes?.[index]?.end && errors?.workTimes?.[index]?.end?.message}
          />
          <Button variant="contained" color="secondary" disabled={isDisabled} onClick={setOpen}>
            DELETE
          </Button>
        </Box>
        <ConfirmDialog isOpen={isOpen} setClose={setClose} body="Can I delete it?" callback={handleRemove} />
      </>
    );
  },
);

Component.displayName = 'Component';
