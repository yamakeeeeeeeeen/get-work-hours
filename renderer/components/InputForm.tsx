import { FC, memo } from 'react';
import { Box, Button, createStyles, makeStyles, Theme, Typography } from '@material-ui/core';
import { Props } from '~/containers/InputForm';
import { TimePickerSet } from '~/containers/TimePickerSet';
import { ResultDialog } from '~/components/ResultDialog';
import { Inputs } from '~/pages';
import { UseFormMethods } from 'react-hook-form';

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    infoText: {
      marginBottom: theme.spacing(3),
    },
  }),
);

export type ComponentProps = Omit<Props, 'append'> & {
  result: { breakTime: string; uptime: string };
  isDisabled: boolean;
  isOpen: boolean;
  setClose: () => void;
  handleAppend: () => void;
  handleConfirm: (val: Inputs) => void;
  handleReset: () => void;
  handleSubmit: UseFormMethods['handleSubmit'];
};

export const Component: FC<ComponentProps> = memo(
  ({ result, isDisabled, isOpen, setClose, handleAppend, handleConfirm, handleReset, handleSubmit, ...props }) => {
    const classes = useStyles();

    return (
      <>
        <Box maxWidth={500} mx="auto">
          <Typography className={classes.infoText}>Please enter the hours you workedüë®‚Äçüíª</Typography>
          {props.fields.map((item, index) => {
            if (item.start && item.end) {
              const value = {
                start: item.start,
                end: item.end,
              };
              return <TimePickerSet key={item.id} index={index} value={value} {...props} />;
            }
            return null;
          })}
          <Box display="flex" justifyContent="space-between">
            <Button variant="contained" color="primary" onClick={handleAppend}>
              APPEND
            </Button>
            <Button variant="contained" color="primary" onClick={handleSubmit(handleConfirm)}>
              CONFIRM
            </Button>
            <Button variant="contained" color="secondary" disabled={!isDisabled} onClick={handleReset}>
              RESET
            </Button>
          </Box>
        </Box>
        <ResultDialog isOpen={isOpen} setClose={setClose} result={result} />
      </>
    );
  },
);

Component.displayName = 'Component';
