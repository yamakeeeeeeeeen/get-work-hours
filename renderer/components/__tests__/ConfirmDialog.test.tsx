import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { ConfirmDialog } from '~/components/ConfirmDialog';

describe('<ConfirmDialog />', () => {
  const body = 'test body';
  const callback = jest.fn();
  const setClose = jest.fn();
  const wrapper = shallow(<ConfirmDialog isOpen={true} body={body} callback={callback} setClose={setClose} />);

  it('snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('The body should be displayed as "test body"', () => {
    expect(wrapper.find('.body-text').text()).toBe(body);
  });

  it('The callback should be called', () => {
    wrapper.find('.apply-btn').simulate('click');
    expect(callback).toHaveBeenCalled();
  });

  it('The setClose should be called', () => {
    wrapper.find('.cancel-btn').simulate('click');
    expect(setClose).toHaveBeenCalled();
  });
});
