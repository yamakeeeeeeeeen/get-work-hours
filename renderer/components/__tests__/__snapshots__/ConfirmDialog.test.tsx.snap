// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConfirmDialog /> snapshot 1`] = `
<WithStyles(ForwardRef(Dialog))
  onClose={[MockFunction]}
  open={true}
>
  <WithStyles(ForwardRef(DialogContent))>
    <WithStyles(ForwardRef(DialogContentText))>
      <WithStyles(ForwardRef(Typography))
        className="body-text"
        variant="h4"
      >
        test body
      </WithStyles(ForwardRef(Typography))>
    </WithStyles(ForwardRef(DialogContentText))>
  </WithStyles(ForwardRef(DialogContent))>
  <WithStyles(ForwardRef(DialogActions))>
    <WithStyles(ForwardRef(Button))
      className="cancel-btn"
      color="secondary"
      onClick={[MockFunction]}
      variant="contained"
    >
      No
    </WithStyles(ForwardRef(Button))>
    <WithStyles(ForwardRef(Button))
      className="apply-btn"
      color="primary"
      onClick={[MockFunction]}
      variant="contained"
    >
      Yes
    </WithStyles(ForwardRef(Button))>
  </WithStyles(ForwardRef(DialogActions))>
</WithStyles(ForwardRef(Dialog))>
`;
